---
title: "Trimming Adapter Sequences From MiRna Reads"
author: "Peter von Rohr"
date: "15 January 2016"
output: pdf_document
---

## Disclaimer
This document summarizes information about how trimming is done for small RNA sequences.

## Current Status
When analysing small RNA sequences with "EzAppNcpro", trimming is currently done using a function called "trimMirna()" which is stored in file "ngsMirna.R". 

In what follows a short summary on trimming is given.

### Function trimMirna
In function "trimMirna()" trimming is done using two programs. 

1. prinseq-lite
2. flexbar

The program `prinseq-lite` runs first by pasting together the following system-command. 

```
gunzip -c /home/petervr/myRepo/ezRun/inst/extdata/smRNA_250k/test1_R1.fastq.gz | \
  /usr/local/ngseq/stow/prinseq-lite-0.20.3/bin/prinseq-lite.pl                  \
    -no_qual_header               \  # empty header line for quality data in fastq files
    -trim_qual_right 20           \  # trim sequences by quality score from 3'-end with given threshold
    -trim_qual_type mean          \  # type of how quality score should be computed
    -trim_qual_window 4           \  # sliding window size used to compute quality score
    -fastq stdin                  \ 
    -out_bad test1_R1_qualtrimBad \
    -out_good test1_R1_qualtrim   \
    > /srv/local/scratch/PVR_test/ncPRO_Result/test1_R1.prinseq.out
```

### Program prinseq-lite
Reference for prinseq-lite is available under http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3051327/. As seen from the above command, prinseq-lite does not do any trimming of the adapter sequence, because the adapter is nowhere specified. What prinseq-lite does, is to filter reads based on a given function or statistic (such as mean, min or max) of the quality score. Which type of statistic should be used is specified by parameter "-trim_qual_type". The statistic is computed on a sliding window the width of which is specified by the parameter "-trim_qual_window". Reads which are trimmed are written to the file indicated by the parameter "-out_bad", whereas the retained reads are written to a file given by the parameter "-out_good". The output caught from STDOUT was empty in all tests done so far.

### Programm flexbar
The following command that is run after prinseq is the following which is using a program called "flexbar". The reference manual for flexbar is available at https://github.com/seqan/flexbar/wiki/Manual. The input of "flexbar" which is read from a file given by the parameter "--reads" corresponds to the reads that were retained after trimming with "prinseq-lite". The output after removing the adapter sequences are written to the file indicated by the option "--target"

```
/usr/local/ngseq/bin/flexbar      \
  --adapter-seq TGGAATTCTCGGGTGCCAAGGAACTCCAGTCAC \
  --adapter-trim-end RIGHT        \   # type of removal, left part of sequence remains
  --adapter-min-overlap 10        \   # minimum overlap between adapter and read
  --adapter-threshold 1.5         \   # allowed gaps and mismatch for removal
  --min-read-length 18            \   # minimum read length after removal for read to stay
  --max-uncalled 4                \   # allowed number of uncalled bases or N's per read
  --format i1.8                   \   # quality format (Illumina 1.8)
  --reads test1_R1_qualtrim.fastq \   # fasta/q files with reads that may contain barcodes/adapters
  --target test1_R1_allTrimmed    \   # prefix for output file names or paths
  > /srv/local/scratch/PVR_test/ncPRO_Result/test1_R1.flexbar.out
```

The output caught from STDOUT prints a summary and some count statistics from the processed read data.


## New Trimming
Trimming as used in function "ncpro()" should be changed to use function "ezMethodTrim()" in "appTrim.R". Parameters of "ezMethodTrim()" are

```
  paired            a logical specifying whether the samples have paired ends.
  subsampleReads    an integer specifying how many subsamples there are. This will call 
                       \code{ezMethodSubsampleReads()} if > 1.
  trimAdapter       a logical specifying whether to use a trim adapter.
  minTailQuality    an integer specifying the minimal tail quality to accept. Only used if > 0.
  minAvgQuality     an integer specifying the minimal average quality to accept. Only used if > 0.
  minReadLength     an integer specifying the minimal read length to accept.
  dataRoot          a character specifying the path of the data root to get the full column paths from.
```

The functions "trimMirna" and "ezMethodTrim()" are not called with the same parameters. The former has parameters "input", "output" and "param" where the first is the input filename, the second is the name of the output file and the last is a list of parameters. The arguments of "ezMethodTrim" have the same name, but are expected to be instances of of the reference class "EzDataset".  

### Function ezMethodTrim()
Trimming in function `ezMethodTrim()` can be done with two programs. 

1. TRIMMOMATIC
2. FLEXBAR

Depending on the specified parameters in list "param", either both, only one of them or none of the trimming programs are run. 

#### Function arguments
Function parameters of function `ezMethodTrim()` are

* input
* output
* param

The first two parameters (input and output) are expected to be instances of the reference class "EzDataset". The function parameter `param` is expected to be a list with components that are listed above as parameters of "ezMethodTrim()". 

#### Output settings
The first step in function "ezMethodTrim()" checks whether an output object of reference class "EzDataset" was passed as function argument, if not output is initialized as a copy of input and paths to read files are appended with the extension "-trimmed-R<n>.fastq" which will later be used as names of result files.

#### Multi-sample Option
The function "ezMethodTrim()" handles multiple samples by calling itself recursively for each single sample given in the input object. 

#### Setting default values
After the recursive calls, the rest of the function is assumed to work with only one sample. The first statements in the single sample section of the function code is setting some default values in the parameter list "param". The components that are initialized are shown in the following table


Component                |  Default value  
------------------------ | ---------------:
trimSeedMismatches       |  1              
trimPalindromClipThresh  |  20              
trimSimpleClipThresh     |  7              
trimMinAdaptLength       |  5              
trimKeepBothReads        |  true           
trimQualWindowWidth      |  4              

#### Read subsampling
In case parameter-components "subsampleReads" > 1 or "nReads" > 0 then the function `ezMethodSubsampleReads()` is called. 

#### Parameters specific for TRIMMOMATIC
The boolean component "trimAdapter" indicates whether specific adapter sequences should be trimmed by TRIMMOMATIC. The parameters given to TRIMMOMATIC are grouped into four different categories which are all saved in a different string variable

1. readOpts: specify input filename and output filename separated by a space
2. trimAdaptOpt: options related to trimming of the adapters. Start with keyword "ILLUMINACLIP", followed by name of file where adapter sequences are stored , and other default options such as: trimSeedMismatches, trimPalindromClipThresh, trimSimpleClipThresh, trimMinAdaptLength, trimKeepBothReads
3. tailQualOpt: tail quality option. Starts with keyword "SLIDINGWINDOW" and contain values of parameter components trimQualWindowWidth and minTailQuality
4. minAvgQualOpt: minimum average quality option. Starts with keyword "AVGQUAL" followed by the parameter component minAvgQuality. 

Adapter sequences are read from a file given by the global variable "TRIMMOMATIC_ADAPTERS". If an adapter sequence is given in the input metainformation, then it is added to the adapters read from the file. The combined set of adapters is used as a command-line argument for TRIMMOMATIC and all adapters are written to a file called "adapters.fa". The name of this file together with other parameter settings related to trimming is written into a colon separated list of options after the keyword "ILLUMINACLIP". The whole string of options is stored in a variable called `trimAdaptOpt`. If the parameter-component "trimAdapter" not set to TRUE, the adapter options in `trimAdaptOpt` are left empty.

#### Parameters specific for FLEXBAR
From the comments in the code, it seams that in the context of ezMethodTrim, unlike when used in function `trimMirna()`, FLEXBAR here is used for fixed trimming, as it is used with option  "--pre-trim-left" followed by a number which most likely indicates the number of positions to be trimmed on the left and similar for option "--pre-trim-right".  


## Comparison
The two trimming routines "trimMirna" und "ezMethodTrim" will be compared by running them on the same dataset and comparing the results after both trimming runs.

These tests are done on the same test data sets as were already used for updating "ncpro". 

### Reference Class EzTrimTester
To make the comparison as easy as possible, a reference class called "EzTrimTester" was created. This does all the preparation and has as runMethod a function that is the same as the one for the ncpro app until the trimming is done. As a result the following files are generated.

```
rw-rw-r-- 1 petervr SG_Users 26292548 Jan 15 16:18 test1_R1.fastq
-rw-rw-r-- 1 petervr SG_Users     1688 Jan 15 16:18 test1_R1.flexbar.out
-rw-rw-r-- 1 petervr SG_Users        0 Jan 15 16:18 test1_R1.prinseq.out
-rw-rw-r-- 1 petervr SG_Users 25323095 Jan 15 16:19 test2_R1.fastq
-rw-rw-r-- 1 petervr SG_Users     1688 Jan 15 16:19 test2_R1.flexbar.out
-rw-rw-r-- 1 petervr SG_Users        0 Jan 15 16:18 test2_R1.prinseq.out
```

where the files with extension .flexbar.out are the output of the flexbar program and the files ending in .fastq are the trimmed readfiles. 

### Results using trimMirna
The result files from running `trimMirna()` are shown above. The number of lines in the fastq files are as follows

```
petervr@fgcz-176:/scratch/PVR_test/ncPRO_Result$ ls -1 *.fastq | while read f;do echo $f;wc -l $f;sleep 2;done
test1_R1.fastq
930856 test1_R1.fastq
test2_R1.fastq
929172 test2_R1.fastq
```

### Results using ezMethodTrim
The results after trimming with TRIMMOMATIC look as follows

```
petervr@fgcz-176:/scratch/PVR_test/ncPRO_Result$ ls -1 *.fastq | while read f;do echo $f;wc -l $f;sleep 2;done
test1-trimmed-R1.fastq
879144 test1-trimmed-R1.fastq
test2-trimmed-R1.fastq
873244 test2-trimmed-R1.fastq
```


## Evaluation of comparison
The two trimming methods

1. `trimMirna` and
2. `ezMethodTrim`

give different results. It would be interesting to check which reads are kept by the first method and whether those are really such bad quality that it is worth to throw them away. 

### Minimum read length
With `trimMirna` the minimum read length is set to 18, when running `ezMethodTrim` with default values it uses minimum read length 20. This brings down the difference to 

for method `ezMethodTrim` 
```
petervr@fgcz-176:/scratch/PVR_test/20160129095934-ezMethodTrim$ ls -1 *.fastq |  while read f;do echo $f;wc -l $f;sleep 2;done
test1-trimmed-R1.fastq
917004 test1-trimmed-R1.fastq
test2-trimmed-R1.fastq
917640 test2-trimmed-R1.fastq
```

and for method `trimMirna`
```
petervr@fgcz-176:/scratch/PVR_test/20160129095934-ezMethodTrim$ ls -1 ../20160129095117-trimmirna/*.fastq |  while read f;do echo $f;wc -l $f;sleep 2;done
../20160129095117-trimmirna/test1_R1.fastq
930856 ../20160129095117-trimmirna/test1_R1.fastq
../20160129095117-trimmirna/test2_R1.fastq
929172 ../20160129095117-trimmirna/test2_R1.fastq
```

### Assessment of read quality
The remaining differences between trimming method were evaluated by checking the quality of the reads that were removed in addition to when using the new trimming method `ezMethodTrim` compared to the existing trimming method `trimMirna`. Read quality was assessed using two different approaches

1. function `qa()` from BioC package `ShortRead` and 
2. the stand-alone tool `fastqc()`

In both cases the output is an html-report which shows different characteristics of the reads.

### Searching for the differentially trimmed reads
Before the quality of the differentially trimmed reads could be assessed, they first had to be identified. With differentially trimmed reads, we mean the reads that were trimmed by one method but not by the other. The results shown here look at the quality of reads that are eliminated by `ezMethodTrim` but not by `trimMirna`. The quality of those reads is interesting, because, if those reads thrown away by `ezMethodTrim` is low, then we can keep the settings and do not have to care about bringing those reads back into any subsequent analyses again.

### Finding differentially trimmed reads
Fastq files that were produced after trimming with both methods are read using function `ShortRead::readFastq()`. This function returns an object of class `ShortReadQ`. Applying the accessor function `id()` returns an object of type `BSString` which can be converted into a vector of read ids using coersion by `as.vector()`. Extracting vector of ids for both fastq files that come out of the trimming process and subsequently computing `setdiff()` on the two vectors of read ids resulted in the set of read ids of differentially trimmed reads. This vector was used to extract the reads from the original `ShortReadQ` object. All of this is implemented in a function called `findReadDiff` The differntially trimmed read are written to a fastq file using `ShortRead::writeFastq()`.  
